File: apps/worker/src/automation_schedule.py
"""Cálculo de próxima execução de automação (timezone-aware).

Espelha apps/web/src/lib/automation-schedule.ts — alterar nos dois lados
juntos. Testes em apps/worker/tests/test_automation_schedule.py.
"""

from __future__ import annotations

from datetime import UTC, datetime, timedelta
from typing import Literal
from zoneinfo import ZoneInfo

Frequency = Literal["DAILY", "WEEKLY", "MONTHLY"]


def compute_next_run(
    *,
    frequency: Frequency,
    hour: int,
    minute: int,
    day_of_week: int | None = None,  # 0=segunda..6=domingo
    day_of_month: int | None = None,  # 1-31 (clampa último dia do mês)
    timezone: str = "America/Sao_Paulo",
    from_dt: datetime | None = None,
) -> datetime:
    """Retorna o próximo datetime UTC em que a regra dispara, estritamente
    depois de `from_dt` (default = agora UTC)."""
    if from_dt is None:
        from_dt = datetime.now(UTC)
    elif from_dt.tzinfo is None:
        from_dt = from_dt.replace(tzinfo=UTC)

    tz = ZoneInfo(timezone)
    local_now = from_dt.astimezone(tz)

    # Constrói candidato local no dia corrente
    candidate_local = local_now.replace(hour=hour, minute=minute, second=0, microsecond=0)
    if candidate_local <= local_now:
        candidate_local += timedelta(days=1)

    if frequency == "DAILY":
        return candidate_local.astimezone(UTC)

    if frequency == "WEEKLY":
        target = day_of_week if day_of_week is not None else 0
        # Python: Monday=0..Sunday=6 — bate com a nossa convenção
        while candidate_local.weekday() != target:
            candidate_local += timedelta(days=1)
        return candidate_local.astimezone(UTC)

    # MONTHLY
    wanted = day_of_month if day_of_month is not None else 1
    # Primeiro tenta no mês corrente
    year = candidate_local.year
    month = candidate_local.month
    attempt_day = _clamp_day_of_month(year, month, wanted)
    candidate_local = candidate_local.replace(
        year=year, month=month, day=attempt_day, hour=hour, minute=minute, second=0, microsecond=0
    )
    if candidate_local <= local_now:
        # Avança pro próximo mês
        if month == 12:
            year += 1
            month = 1
        else:
            month += 1
        attempt_day = _clamp_day_of_month(year, month, wanted)
        candidate_local = candidate_local.replace(year=year, month=month, day=attempt_day)
    return candidate_local.astimezone(UTC)


def _clamp_day_of_month(year: int, month: int, wanted: int) -> int:
    """Retorna o `wanted` clampado para o último dia válido do mês."""
    # Último dia: vai pro próximo mês e volta um dia
    if month == 12:
        next_month_first = datetime(year + 1, 1, 1)
    else:
        next_month_first = datetime(year, month + 1, 1)
    last_day = (next_month_first - timedelta(days=1)).day
    return max(1, min(wanted, last_day))

File: apps/web/src/lib/automation-schedule.ts
// ============================================================================
// Cálculo de próxima execução de automação (timezone-aware).
// ============================================================================
// Espelha a lógica em apps/worker/src/automation_schedule.py — alterar
// nos dois lados juntos. Testes em tests/automation-schedule.test.ts.
// ============================================================================

export type Frequency = 'DAILY' | 'WEEKLY' | 'MONTHLY';

export interface ScheduleInput {
  frequency: Frequency;
  hour: number; // 0-23 no timezone
  minute: number; // 0-59
  dayOfWeek?: number | null; // 0=segunda..6=domingo (WEEKLY)
  dayOfMonth?: number | null; // 1-31 (MONTHLY, clampa último dia)
  timezone: string; // IANA, ex: "America/Sao_Paulo"
}

// Retorna o próximo Date UTC em que a regra dispara, estritamente depois
// de `from`. Estratégia: trabalha no calendário do timezone do user
// avançando o "candidato" até casar a regra, depois converte pra UTC via
// Intl.DateTimeFormat (idempotente em DST porque sempre vamos pra frente
// — não há "duas vezes na mesma hora").
export function computeNextRun(input: ScheduleInput, from: Date = new Date()): Date {
  const { frequency, hour, minute, timezone } = input;
  // Parte do "hoje no tz do user" às HH:MM
  const local = toLocalParts(from, timezone);
  let year = local.year;
  let month = local.month; // 1-12
  let day = local.day;

  // Constrói candidato no calendário local
  let candidate = makeUtcFromLocal(year, month, day, hour, minute, timezone);
  if (candidate <= from) {
    // Já passou hoje — avança 1 dia
    const tomorrow = addDaysLocal(year, month, day, 1);
    year = tomorrow.year;
    month = tomorrow.month;
    day = tomorrow.day;
    candidate = makeUtcFromLocal(year, month, day, hour, minute, timezone);
  }

  if (frequency === 'DAILY') {
    return candidate;
  }

  if (frequency === 'WEEKLY') {
    const target = input.dayOfWeek ?? 0; // 0=segunda..6=domingo
    while (localDayOfWeekMonStart(candidate, timezone) !== target) {
      const next = addDaysLocal(year, month, day, 1);
      year = next.year;
      month = next.month;
      day = next.day;
      candidate = makeUtcFromLocal(year, month, day, hour, minute, timezone);
    }
    return candidate;
  }

  // MONTHLY — pega o dia desejado, clampa pro último do mês se exceder
  const wanted = input.dayOfMonth ?? 1;
  // Primeira tentativa: dia desejado no mês corrente
  let attemptDay = clampDayOfMonth(year, month, wanted);
  candidate = makeUtcFromLocal(year, month, attemptDay, hour, minute, timezone);
  if (candidate <= from) {
    // Já passou esse mês — vai pro próximo
    if (month === 12) {
      year += 1;
      month = 1;
    } else {
      month += 1;
    }
    attemptDay = clampDayOfMonth(year, month, wanted);
    candidate = makeUtcFromLocal(year, month, attemptDay, hour, minute, timezone);
  }
  return candidate;
}

// ---------------------------------------------------------------------------
// Helpers — manipulação de calendário local sem libs externas.
// ---------------------------------------------------------------------------

interface LocalParts {
  year: number;
  month: number; // 1-12
  day: number; // 1-31
  hour: number;
  minute: number;
}

function toLocalParts(date: Date, timezone: string): LocalParts {
  const fmt = new Intl.DateTimeFormat('en-CA', {
    timeZone: timezone,
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: '2-digit',
    minute: '2-digit',
    hour12: false,
  });
  const parts = fmt.formatToParts(date);
  const get = (t: string): number => Number(parts.find((p) => p.type === t)?.value ?? '0');
  // hourCycle h23 (00-23) — em alguns runtimes vem '24'; normaliza
  let h = get('hour');
  if (h === 24) h = 0;
  return {
    year: get('year'),
    month: get('month'),
    day: get('day'),
    hour: h,
    minute: get('minute'),
  };
}

// Cria um Date UTC tal que, no timezone informado, o calendário mostre
// year/month/day/hour/minute. Usa busca iterativa com o offset do tz
// (suficientemente precisa: convergente em 1-2 iterações).
function makeUtcFromLocal(
  year: number,
  month: number,
  day: number,
  hour: number,
  minute: number,
  timezone: string,
): Date {
  // Chute inicial: trata como se fosse UTC
  let candidate = new Date(Date.UTC(year, month - 1, day, hour, minute, 0, 0));
  for (let i = 0; i < 3; i++) {
    const parts = toLocalParts(candidate, timezone);
    const diffMinutes =
      (year - parts.year) * 525_600 +
      (month - parts.month) * 43_200 + // aprox; corrige no próximo passo
      (day - parts.day) * 1440 +
      (hour - parts.hour) * 60 +
      (minute - parts.minute);
    if (diffMinutes === 0) return candidate;
    candidate = new Date(candidate.getTime() + diffMinutes * 60_000);
  }
  return candidate;
}

function addDaysLocal(
  year: number,
  month: number,
  day: number,
  delta: number,
): { year: number; month: number; day: number } {
  // Usa Date para somar dias (em UTC, mas só nos importamos com o calendário)
  const d = new Date(Date.UTC(year, month - 1, day, 12, 0, 0));
  d.setUTCDate(d.getUTCDate() + delta);
  return { year: d.getUTCFullYear(), month: d.getUTCMonth() + 1, day: d.getUTCDate() };
}

function clampDayOfMonth(year: number, month: number, wanted: number): number {
  const lastDay = new Date(Date.UTC(year, month, 0)).getUTCDate(); // month aqui é 1-12 → próximo mês dia 0 = último do mês corrente
  return Math.min(Math.max(1, wanted), lastDay);
}

// Retorna dia da semana 0=segunda..6=domingo no tz informado.
function localDayOfWeekMonStart(date: Date, timezone: string): number {
  const fmt = new Intl.DateTimeFormat('en-US', {
    timeZone: timezone,
    weekday: 'short',
  });
  const wk = fmt.format(date);
  const map: Record<string, number> = { Mon: 0, Tue: 1, Wed: 2, Thu: 3, Fri: 4, Sat: 5, Sun: 6 };
  return map[wk] ?? 0;
}
